include/aws/common/task_scheduler.h

```diff
@@ -0,0 +1,101 @@
+#ifndef AWS_TASK_SCHEDULER_H
+#define AWS_TASK_SCHEDULER_H
+
+/*
+ * Copyright 2010-2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/common.h>
+#include <aws/common/priority_queue.h>
+#include <stdint.h>
+
+typedef void *aws_task_arg;
+
+/**
+ * Pointer to the scheduled function
+ */
+typedef void(*aws_task_fn)(aws_task_arg arg);
+
+/**
+ * Monotonic clock function
+ */
+typedef int(*aws_task_scheduler_clock)(uint64_t *);
+
+struct aws_task {
+    aws_task_fn fn;
+    aws_task_arg arg;
+};
+
+struct aws_task_scheduler {
+    struct aws_allocator *alloc;
+    struct aws_priority_queue queue;
+    uint64_t owning_thread;
+    aws_task_scheduler_clock clock;
+};
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+    /**
+     * Initializes a task scheduler instance with a clock function and config. A priority queue will be allocated 
+     * in dynamic mode with initial size of 7
+     * The recommended clock functions are in <aws/common/clock.h>
+     *
+     * Note: This needs to be called from the thread that will own the scheduler. Cross-thread access is not allowed 
+     * and will always result in an error code being returned.
```

Comment:
Don't commit to actively detecting illegal x-thread access and reporting it. This can be expensive, depending on the platform.

Furthermore, _why_ is it illegal to move a task scheduler around? Couldn't I use it safely on multiple threads if I added appropriate locking around it? Should all the other datastructures guard against cross-thread access too?

Reply:
The task scheduler design is exclusively for single-threaded scenarios, and where threads are being crossed, we want something like an event loop to do the messaging. This will make broken code look broken at debug and development time. 

Arguably we could forgo this and do a wrapper in aws-c-io, that does the check, but some other products at Amazon passed on their wisdom to me from some COEs that asserting or erroring on an accidental thread jump will keep it from accidentally occurring.